evm/contracts/examples/ConversionRate.sol

```diff
@@ -177,16 +177,61 @@ contract ConversionRate is ChainlinkClient, Ownable {
     uint256 sumQuotients;
     uint256 sumRemainders;
     Answer memory answer = answers[_answerId];
-    uint256 responseCount = answer.responses.length;
-    for (uint i = 0; i < responseCount; i++) {
-      uint256 response = answer.responses[i];
-      sumQuotients = sumQuotients.add(response.div(responseCount)); // aggregate responses and protect from overflows
-      sumRemainders = sumRemainders.add(response % responseCount);
+
+    uint256 responseLength = answer.responses.length;
+    uint256 middleIndex = responseLength / 2;
+    if (responseLength % 2 == 0) {
+      uint256 median1 = answers[_answerId].responses[middleIndex];
+      uint256 median2 = answers[_answerId].responses[middleIndex - 1];
+      currentRate = median1.add(median2).div(2);
+    } else {
+      currentRate = answers[_answerId].responses[middleIndex];
     }
-    currentRate = sumQuotients.add(sumRemainders.div(responseCount)); // recover lost accuracy from result 
+
     latestCompletedAnswer = _answerId;
   }
 
+  function insertResponse(uint256 _id, uint256 _response)
+    private
+  {
+    uint256 responseLength = answers[_id].responses.length;
+    answers[_id].responses.length++;
+    uint256 index = findInsertionIndex(_id, _response, responseLength);
```

Comment:
I think it would be better if we just sorted when it's time to provide an answer. The gas usage on this will grow enormous as the consensus set gets bigger. Another possibility would be to have a second method which is called to calculate and update the rate, once enough reports are available. That method could take the median as input, and verify it with a simple, linear count of the array.